// express-checkout/lib/components/PaymentsSummaryContainer/redux.js
"use strict";

exports.__esModule = true;
exports.confirmAccount = exports.storeConfirmAccount = exports.confirmAccountError = exports.setConfirmAccountError = exports.confirmAccountLoading = exports.setConfirmAccountLoading = exports.presetAccount = exports.storePresetAccount = exports.presetAccountError = exports.setPresetAccountError = exports.presetAccountLoading = exports.setPresetAccountLoading = undefined;

var _createReducer, _createReducer2, _createReducer3, _createReducer4, _createReducer5, _createReducer6;

var _toolkit = require("@reduxjs/toolkit");

var initialBooleanState = false;
var initialObjectState = {};

var setPresetAccountLoading = (0, _toolkit.createAction)("PRESETACCOUNTLOADING");
var presetAccountLoading = (0, _toolkit.createReducer)(initialBooleanState, (_createReducer = {}, _createReducer[setPresetAccountLoading] = function (state, action) {
    return action.payload;
}, _createReducer));

var setPresetAccountError = (0, _toolkit.createAction)("PRESETACCOUNTERROR");
var presetAccountError = (0, _toolkit.createReducer)(initialObjectState, (_createReducer2 = {}, _createReducer2[setPresetAccountError] = function (state, action) {
    return action.payload;
}, _createReducer2));

var storePresetAccount = (0, _toolkit.createAction)("PRESETACCOUNT");
var presetAccount = (0, _toolkit.createReducer)(initialObjectState, (_createReducer3 = {}, _createReducer3[storePresetAccount] = function (state, action) {
    return action.payload;
}, _createReducer3));

var setConfirmAccountLoading = (0, _toolkit.createAction)("CONFIRMACCOUNTLOADING");
var confirmAccountLoading = (0, _toolkit.createReducer)(initialBooleanState, (_createReducer4 = {}, _createReducer4[setConfirmAccountLoading] = function (state, action) {
    return action.payload;
}, _createReducer4));

var setConfirmAccountError = (0, _toolkit.createAction)("CONFIRMACCOUNTERROR");
var confirmAccountError = (0, _toolkit.createReducer)(initialObjectState, (_createReducer5 = {}, _createReducer5[setConfirmAccountError] = function (state, action) {
    return action.payload;
}, _createReducer5));

var storeConfirmAccount = (0, _toolkit.createAction)("CONFIRMACCOUNT");
var confirmAccount = (0, _toolkit.createReducer)(initialObjectState, (_createReducer6 = {}, _createReducer6[storeConfirmAccount] = function (state, action) {
    return action.payload;
}, _createReducer6));

exports.setPresetAccountLoading = setPresetAccountLoading;
exports.presetAccountLoading = presetAccountLoading;
exports.setPresetAccountError = setPresetAccountError;
exports.presetAccountError = presetAccountError;
exports.storePresetAccount = storePresetAccount;
exports.presetAccount = presetAccount;
exports.setConfirmAccountLoading = setConfirmAccountLoading;
exports.confirmAccountLoading = confirmAccountLoading;
exports.setConfirmAccountError = setConfirmAccountError;
exports.confirmAccountError = confirmAccountError;
exports.storeConfirmAccount = storeConfirmAccount;
exports.confirmAccount = confirmAccount;
