// express-checkout/lib/index.js
'use strict';

exports.__esModule = true;

require('react-app-polyfill/ie11');

require('react-app-polyfill/stable');

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactRedux = require('react-redux');

var _store = require('./store');

var _store2 = _interopRequireDefault(_store);

var _ConfigurationManager = require('./components/ConfigurationManager');

var _ConfigurationManager2 = _interopRequireDefault(_ConfigurationManager);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * The main component to render express checkout widget
 * @param {Object} props
 * @param {Object} props.configuration
 * @param {Function} props.createTransactionDetails
 * @param {Object} props.customFunctions
 * @param {String} props.mode
 * @param {String} props.longId
 * @return {JSX.Element}
 */
var ExpressCheckout = function ExpressCheckout(_ref) {
    var configuration = _ref.configuration,
        createTransactionDetails = _ref.createTransactionDetails,
        customFunctions = _ref.customFunctions,
        mode = _ref.mode,
        longId = _ref.longId;

    var store = (0, _store2.default)();
    return _react2.default.createElement(
        _reactRedux.Provider,
        { store: store },
        _react2.default.createElement(_ConfigurationManager2.default, {
            configuration: configuration,
            createTransactionDetails: createTransactionDetails,
            customFunctions: customFunctions,
            mode: mode,
            longId: longId
        })
    );
};

ExpressCheckout.propTypes = process.env.NODE_ENV !== "production" ? {
    configuration: _propTypes2.default.object.isRequired,
    createTransactionDetails: _propTypes2.default.func.isRequired,
    customFunctions: _propTypes2.default.object,
    mode: _propTypes2.default.string,
    longId: _propTypes2.default.string
} : {};

exports.default = ExpressCheckout;
module.exports = exports['default'];
