// express-checkout/es/components/ConfigurationManager/redux.js
var _createReducer, _createReducer2, _createReducer3;

import { createReducer, createAction } from "@reduxjs/toolkit";
/**
 * initial empty string
 */
var initialStringState = "";
var initialConfigurationState = {
    baseURL: "",
    clientId: "",
    country: "",
    paymentMethodsConfiguration: []
};

var storeMode = createAction("STOREMODE");
var mode = createReducer(initialStringState, (_createReducer = {}, _createReducer[storeMode] = function (state, action) {
    return action.payload;
}, _createReducer));

var storeLongId = createAction("STORELONGID");
var longId = createReducer(initialStringState, (_createReducer2 = {}, _createReducer2[storeLongId] = function (state, action) {
    return action.payload;
}, _createReducer2));

var storeConfiguration = createAction("STORECONFIGURATION");
var configuration = createReducer(initialConfigurationState, (_createReducer3 = {}, _createReducer3[storeConfiguration] = function (state, action) {
    return action.payload;
}, _createReducer3));

export { mode, storeMode, longId, storeLongId, storeConfiguration, configuration };
