// express-checkout/lib/store/reducer.js
"use strict";

exports.__esModule = true;

var _redux = require("redux");

var _redux2 = require("../components/ConfigurationManager/redux");

var _redux3 = require("../components/PaymentsContainer/redux");

var _redux4 = require("../components/GlobalError/redux");

var _redux5 = require("../components/PaymentsSummaryContainer/redux");

var _redux6 = require("../components/paymentMethods/Paypal/redux");

var rootReducer = (0, _redux.combineReducers)({
    mode: _redux2.mode,
    longId: _redux2.longId,
    configuration: _redux2.configuration,
    list: (0, _redux.combineReducers)({
        data: _redux3.list,
        listError: _redux3.listError,
        listLoading: _redux3.listLoading
    }),
    paypal: (0, _redux.combineReducers)({
        paypalStatus: _redux6.paypalStatus,
        paymentID: _redux6.paymentID,
        preset: _redux6.preset,
        cancelData: _redux6.cancelData,
        error: _redux6.error
    }),
    presetAccount: (0, _redux.combineReducers)({
        data: _redux5.presetAccount,
        presetAccountError: _redux5.presetAccountError,
        presetAccountLoading: _redux5.presetAccountLoading
    }),
    confirmAccount: (0, _redux.combineReducers)({
        data: _redux5.confirmAccount,
        confirmAccountError: _redux5.confirmAccountError,
        confirmAccountLoading: _redux5.confirmAccountLoading
    }),
    globalError: _redux4.globalError,
    displayGlobalError: _redux4.displayGlobalError
});

exports.default = rootReducer;
module.exports = exports["default"];
