// express-checkout/es/store/reducer.js
import { combineReducers } from "redux";

import { configuration, mode, longId } from "../components/ConfigurationManager/redux";
import { list, listError, listLoading } from "../components/PaymentsContainer/redux";
import { globalError, displayGlobalError } from "../components/GlobalError/redux";
import { presetAccount, presetAccountError, presetAccountLoading, confirmAccountLoading, confirmAccountError, confirmAccount } from "../components/PaymentsSummaryContainer/redux";
import { paypalStatus, paymentID, preset, cancelData, error } from "../components/paymentMethods/Paypal/redux";

var rootReducer = combineReducers({
    mode: mode,
    longId: longId,
    configuration: configuration,
    list: combineReducers({
        data: list,
        listError: listError,
        listLoading: listLoading
    }),
    paypal: combineReducers({
        paypalStatus: paypalStatus,
        paymentID: paymentID,
        preset: preset,
        cancelData: cancelData,
        error: error
    }),
    presetAccount: combineReducers({
        data: presetAccount,
        presetAccountError: presetAccountError,
        presetAccountLoading: presetAccountLoading
    }),
    confirmAccount: combineReducers({
        data: confirmAccount,
        confirmAccountError: confirmAccountError,
        confirmAccountLoading: confirmAccountLoading
    }),
    globalError: globalError,
    displayGlobalError: displayGlobalError
});

export default rootReducer;
