// express-checkout/lib/components/paymentMethods/Paypal/redux.js
"use strict";

exports.__esModule = true;
exports.storePaypalError = exports.storePaypalCancelData = exports.storePaypalPreset = exports.storePaypalPaymentID = exports.storePaypalStatus = exports.error = exports.cancelData = exports.preset = exports.paymentID = exports.paypalStatus = undefined;

var _createReducer, _createReducer2, _createReducer3, _createReducer4, _createReducer5;

var _toolkit = require("@reduxjs/toolkit");

var initialStringState = "";
var initialObjectState = {};

var storePaypalStatus = (0, _toolkit.createAction)("STOREPAYPALSTATUS");
var paypalStatus = (0, _toolkit.createReducer)(initialStringState, (_createReducer = {}, _createReducer[storePaypalStatus] = function (state, action) {
    return action.payload;
}, _createReducer));
var storePaypalPaymentID = (0, _toolkit.createAction)("STOREPAYPALPAYMENTID");
var paymentID = (0, _toolkit.createReducer)(initialStringState, (_createReducer2 = {}, _createReducer2[storePaypalPaymentID] = function (state, action) {
    return action.payload;
}, _createReducer2));
var storePaypalPreset = (0, _toolkit.createAction)("STOREPAYPALPRESET");
var preset = (0, _toolkit.createReducer)(initialObjectState, (_createReducer3 = {}, _createReducer3[storePaypalPreset] = function (state, action) {
    return action.payload;
}, _createReducer3));
var storePaypalCancelData = (0, _toolkit.createAction)("STOREPAYPALCANCELDATA");
var cancelData = (0, _toolkit.createReducer)(initialObjectState, (_createReducer4 = {}, _createReducer4[storePaypalCancelData] = function (state, action) {
    return action.payload;
}, _createReducer4));
var storePaypalError = (0, _toolkit.createAction)("STOREPAYPALERROR");
var error = (0, _toolkit.createReducer)(initialObjectState, (_createReducer5 = {}, _createReducer5[storePaypalError] = function (state, action) {
    return action.payload;
}, _createReducer5));

exports.paypalStatus = paypalStatus;
exports.paymentID = paymentID;
exports.preset = preset;
exports.cancelData = cancelData;
exports.error = error;
exports.storePaypalStatus = storePaypalStatus;
exports.storePaypalPaymentID = storePaypalPaymentID;
exports.storePaypalPreset = storePaypalPreset;
exports.storePaypalCancelData = storePaypalCancelData;
exports.storePaypalError = storePaypalError;
