// express-checkout/lib/components/PaymentsContainer/redux.js
"use strict";

exports.__esModule = true;
exports.list = exports.storeList = exports.listError = exports.setListError = exports.listLoading = exports.setListLoading = undefined;

var _createReducer, _createReducer2, _createReducer3;

var _toolkit = require("@reduxjs/toolkit");

var initialBooleanState = false;
var initialObjectState = {};
var initialArrayState = [];

var setListLoading = (0, _toolkit.createAction)("LISTLOADING");
var listLoading = (0, _toolkit.createReducer)(initialBooleanState, (_createReducer = {}, _createReducer[setListLoading] = function (state, action) {
    return action.payload;
}, _createReducer));

var setListError = (0, _toolkit.createAction)("LISTERROR");
var listError = (0, _toolkit.createReducer)(initialObjectState, (_createReducer2 = {}, _createReducer2[setListError] = function (state, action) {
    return action.payload;
}, _createReducer2));

var storeList = (0, _toolkit.createAction)("STORELIST");
var list = (0, _toolkit.createReducer)(initialArrayState, (_createReducer3 = {}, _createReducer3[storeList] = function (state, action) {
    return action.payload;
}, _createReducer3));

exports.setListLoading = setListLoading;
exports.listLoading = listLoading;
exports.setListError = setListError;
exports.listError = listError;
exports.storeList = storeList;
exports.list = list;
