// express-checkout/lib/components/ConfigurationManager/redux.js
"use strict";

exports.__esModule = true;
exports.configuration = exports.storeConfiguration = exports.storeLongId = exports.longId = exports.storeMode = exports.mode = undefined;

var _createReducer, _createReducer2, _createReducer3;

var _toolkit = require("@reduxjs/toolkit");

/**
 * initial empty string
 */
var initialStringState = "";
var initialConfigurationState = {
    baseURL: "",
    clientId: "",
    country: "",
    paymentMethodsConfiguration: []
};

var storeMode = (0, _toolkit.createAction)("STOREMODE");
var mode = (0, _toolkit.createReducer)(initialStringState, (_createReducer = {}, _createReducer[storeMode] = function (state, action) {
    return action.payload;
}, _createReducer));

var storeLongId = (0, _toolkit.createAction)("STORELONGID");
var longId = (0, _toolkit.createReducer)(initialStringState, (_createReducer2 = {}, _createReducer2[storeLongId] = function (state, action) {
    return action.payload;
}, _createReducer2));

var storeConfiguration = (0, _toolkit.createAction)("STORECONFIGURATION");
var configuration = (0, _toolkit.createReducer)(initialConfigurationState, (_createReducer3 = {}, _createReducer3[storeConfiguration] = function (state, action) {
    return action.payload;
}, _createReducer3));

exports.mode = mode;
exports.storeMode = storeMode;
exports.longId = longId;
exports.storeLongId = storeLongId;
exports.storeConfiguration = storeConfiguration;
exports.configuration = configuration;
